Add support for Ubuntu Bionic for StackStorm using Python 3

Kami opened the pull request:
This pull request add support for building StackStorm packages for Ubuntu Bionic using Python 3.

## Depends on

* https://github.com/StackStorm/st2packaging-dockerfiles/pull/69

## Related PRs

* https://github.com/StackStorm/st2web/pull/624
* https://github.com/StackStorm/st2chatops/pull/117
* https://github.com/StackStorm/bwc-enterprise-packages/pull/26

## TODO

- [x] New dh virtualenv debian rule for Bionic which uses python 3 binary
- [x] Support for Bionic in installer script (only for staging unstable until it's officially supported)

armab commented on README.md:
```diff
@@ -107,6 +105,67 @@ Current community packages are hosted on https://packagecloud.io/StackStorm. For
+images are not yet published to Docker Hub and you want to test the build pipeline locally, you
+need to update ``docker-compose.yml`` file to use locally built Docker images.
+
+For example:
```
Please include references to files and high-level ideas abut what needs to be done, but not the code samples as it's gets outdated quickly and every distro could have their own specifics.

Kami replied:
It's not meant to be copied directly, just to give user some general idea on what needs to be done.

Kami replied:
Will take a look at it again tomorrow and re-think / re-word it if I can, worst case, I will just remove it.

My goal was to just add something which helped me and I thought it may perhaps or also add others.

armab commented on scripts/st2bootstrap-deb.template.sh:
```diff
@@ -74,6 +75,14 @@ setup_args() {
     fi
   fi
 
+  # Right now Bionic is not officially supported yet so we only support using staging unstable packages
```
Works to me :+1:

Kami commented on docker-compose.circle.yml:
```diff
@@ -22,6 +22,26 @@ xenial:
+    # NOTE: We don't build st2mistral package so we don't mix Python 3
+    # (StackStorm components) and Python 2 (Mistral)
+    # Bionic will be first release without a Mistral
+    - ST2_GITREV=fix_py3_issues
```
Yeah, the plan was to get that one merged before this PR.

Kami commented on packages/st2/Makefile:
```diff
@@ -4,10 +4,13 @@ ST2PKG_RELEASE ?= 1
 ST2PKG_VERSION ?= $(shell python -c "from $(ST2_COMPONENT) import __version__; print __version__,")
 CHANGELOG_COMMENT ?= "automated build, version: $(ST2PKG_VERSION)"
 RUNNERS := $(shell ls ../contrib/runners)
+DEB_DISTRO := $(shell lsb_release -cs)
```
Good catch, will update it.

Kami commented:
Re CI issues and packaging not being produced yet - yeah, they should get produced once this PR is merged (https://github.com/StackStorm/st2packaging-dockerfiles/pull/69).

Will follow it through tomorrow once that PR is merged. In theory, it should work out of the box, might just need to tag a release in that repository, similar to what it looks like we do for other distros and then I assume Docker Hub should automatically build the image and tag it appropriately.

armab commented on docker-compose.circle.yml:
```diff
@@ -22,6 +22,25 @@ xenial:
+    # NOTE: We don't build st2mistral package so we don't mix Python 3
+    # (StackStorm components) and Python 2 (Mistral)
+    # Bionic will be first release without a Mistral
+    - ST2_PACKAGES=st2
```
Try to set an `ST2_PACKAGES` ENV in docker-compose definition for both `bionicbuild` and `bionictest` (instead of `buildenv_st2.sh`)

Kami replied:
Thanks - I was actually thinking about trying that, but then I checked at the code and assumed buildenv stuff still has priority based on the way how it's passed to docker compose.

Kami replied:
It looks like it indeed doesn't work - https://circleci.com/gh/StackStorm/st2-packages/3180#queue-placeholder/containers/2 and I will need to go with my original approach.

As I mentioned above, I wanted to try this first then I checked docker compose script and I assumed it won't work because of the way strict is structured.

armab replied:
And so we use the one taking most precedence via `-e` :/

Kami commented:
Alright, the Bionic packages are building now.

I needed to make various changes to make it work:

1. Change Circle CI parallelism from 4 to 5 so we build packages for all the distros (once we drop support for Trusty we will be back to 4 again).
2. Update `packagecloud.sh` script so it supports Bionic. Right now we only upload Bionic packages for staging unstable (intentional, until we officially announce support for it).
3. Update `buildenv_st2.sh` script so it correctly sets `ST2_PACKAGES` environment variable so we only build st2 package and not mistral on Bionic. @armab and I had a discussion on this above. Sadly due to the way how our `docker-compose2.sh` script is structured, I can't just specify those values in docker compose definition file, but need to utilize this script so it works correctly on Circle CI (small discrepancy between local and Circle CI build environment).
4. Various other changes.

Next step is testing and getting bootstrap script to work on Bionic. One that works, I think we should review this and merge it into master.

And after that (or perhaps before), I will deal with Docker Hub repo recreation drama.

armab commented on .circle/buildenv_st2.sh:
```diff
@@ -51,6 +58,13 @@ re="\\b$DISTRO\\b"
+if [ "${DISTRO}" = "bionic"} ]; then
+    ST2_PACKAGES="st2"
+else
+    ST2_PACKAGES=${ST2_PACKAGES:-st2 mistral}
```
`buildenv_st2.sh` is intended to be st2-focused only, while `buildenv_mistral.sh` is about mistral only and so they're responsible for different packages but not alltogether.

This part I definitely don't like.

Will try different things to workaround it, - but in a meantime we have `buildenv_common.sh` which can be favored instead.

Kami replied:
Yeah, I'm also not a big fan of it either. For now, I will move it into ``buildenv_common.sh``.

Maybe eventually we can refactor how our docker-compose2.sh script works so it will only pass environment variable to docker compose if it's set.

Kami commented on .circle/buildenv_st2.sh:
```diff
@@ -1,21 +1,27 @@
 #!/bin/bash
 set -e
+set -x
```
Yeah, although the default output is very unfriendly - if there is an issue in this script it will just show "script exited with 1" which provides no helpful / debug info on failure.

armab replied:
We really touch this part of st2-packages once in a year.

With that much verbosity what's more important like "export" statements with actual values, - they become less visible and harder to quick-read.
This noise makes it harder to see what's important in a day-to-day builds.

Alternative is to improve script error reporting for the edge cases, instead of using debug mode `-x`.

Kami replied:
I will remove it for now since I don't have the motivation to re-do all the error handling in that script.

Kami commented on .circle/buildenv_st2.sh:
```diff
@@ -1,21 +1,27 @@
+echo "Using distro: ${DISTRO}"
+echo "Using Python: $(python --version 2>&1)"
+echo "ST2_GITURL: ${ST2_GITURL}"
+echo "ST2_GITREV: ${ST2_GITREV}"
```
Yep, but I wanted to print it before if some step in the script fails (which happened in my case and see comment below - default output is useless and provides no info which would help with debugging).

armab replied:
But we're done with the debugging stage, right?

Kami replied:
I will remove last two, but leave Python version and distro in place since it can still come handy.

armab commented on packages/st2/debian/control:
```diff
@@ -4,16 +4,16 @@ Priority: optional
 
 Package: st2
 Architecture: any
-Pre-Depends: dpkg (>= 1.16.16), python2.7, ${misc:Pre-Depends}, adduser
```
I wouldn't touch such things like `${misc:Pre-Depends}` as it's something that's autogenerated by the debhelper based on dependencies he'll find package relies on.

Kami replied:
Yeah, I was just testing it - I will put it back once I confirm everything is working.

armab commented on packages/st2/debian/rules:
```diff
@@ -13,16 +13,41 @@ DH_VIRTUALENV_INSTALL_ROOT := /opt/stackstorm
+OTHER_DEPENDS = libpython-dev | python-dev
+
+ifeq ($(DEB_DISTRO),bionic)
+	BUILD_PRE_DEPENDS := $(BIONIC_PRE_DEPENDS)
```
Because `BIONIC_PRE_DEPENDS` and similar vars are not used everywhere more than 1 time, can we just simplify it to something like

```suggestion
	BUILD_PRE_DEPENDS := python3 (>= 3.6)
```

This will move real values closer to context, making it easier to understand the flow.
Otherwise var->var->var dependency is too Pythonic :)

armab commented on packages/st2/debian/control:
```diff
@@ -4,16 +4,16 @@ Priority: optional
-Pre-Depends: dpkg (>= 1.16.16), python2.7, ${misc:Pre-Depends}, adduser
-Depends: ${shlibs:Depends}, ${misc:Depends}, sudo, adduser, libpython-dev | python-dev, libssl-dev, libffi-dev, git, libpam0g, openssh-server, openssh-client, bash, netbase
+Pre-Depends: dpkg (>= 1.16.16), ${pre:Depends}, ${misc:Pre-Depends}, adduser
+Depends: ${shlibs:Depends}, ${misc:Depends}, sudo, adduser, ${Depends}, libssl-dev, libffi-dev, git, libpam0g, openssh-server, openssh-client, bash, netbase
```
:+1:

armab commented on .circleci/config.yml:
```diff
@@ -122,6 +121,7 @@ jobs:
       - image: circleci/ruby:2.4
     working_directory: /tmp/deploy
     environment:
+      # TODO: Add support for deploying Bionic to Staging Unstable
```
As I understand this could be removed and bionic is OK to add to the deployment list as you already covered the Packagecloud.sh changes

Kami replied:
Correct - will do.

armab commented on .circle/buildenv_st2.sh:
```diff
@@ -51,6 +55,13 @@ re="\\b$DISTRO\\b"
 ST2_CIRCLE_URL=${CIRCLE_BUILD_URL}
 
-write_env ST2_GITURL ST2_GITREV ST2PKG_VERSION ST2PKG_RELEASE DISTRO TESTING ST2_CIRCLE_URL
+# NOTE: We don't build Mistral package on Ubuntu Bionic
```
As code was moved to `buildenv_common.sh` now and can revert injecting this logic in `buildenv_st2.sh`?

Kami replied:
Done.

Kami commented:
@armab I believe I addressed all the comment and this PR is now ready for a final review.

I think it would be good to merge st2web, st2chatops and st2enterprise changes to those Ubuntu Bionic packages are built and uploaded to packagecloud.io.

This way I can test Bionic installer script end to end - basically I tested everything except st2web and st2chatops and enterprise for which we don't have packages at the moment so I can't test that.

Kami commented:
For the approach mentioned above to work, I will need to merge ``packagecloud.sh`` changes into master before other changes in this branch.

I will create a separate PR with changes in https://github.com/StackStorm/st2-packages/pull/590/files#diff-c1a0d8b9f26079f47f9a3565b51a2f89R1 which can be merged immediately (it's fully backward and forward compatible, has no negative implications).

armab commented on scripts/st2bootstrap-deb.template.sh:
```diff
@@ -114,14 +128,17 @@ install_st2_dependencies() {
   # Note: gnupg-curl is needed to be able to use https transport when fetching keys
-  sudo apt-get install -y gnupg-curl
+  if [[ "$SUBTYPE" != 'bionic' ]]; then
+    sudo apt-get install -y gnupg-curl
```
This specific package doesn't present in Bionic?

Kami replied:
Yeah, it's not (https://packages.ubuntu.com/search?keywords=gnupg-curl) and https transport for curl adding keys to APT works by default.

Kami commented:
Now that all the other repo (st2chatops, st2web, bwc-enterprise) PRs have been merged, I confirmed this works end to end:

1. Installer script works
2. self check works
3. Basic functionality works

I will go ahead and merge this into master and start working on hooking this up to cicd.
